sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -579,36 +581,66 @@ static inline int find_codec_info_acpi(const u8 *acpi_id)
 	return i;
 }
 
-static int get_sdw_dai_link_num(struct snd_soc_acpi_mach_params *mach_params)
+/*
+ * get BE DAI and CPU DAI number based on sdw link adr.
+ * Since some sdw links may be aggregated to support one
+ * PCM, CPU DAI number is not always equal to BE DAI.
+ */
+static int get_sdw_dai_info(struct snd_soc_acpi_mach_params *mach_params,
+			    int *sdw_dai_num, int *cpu_dai_num)
 {
 	const struct snd_soc_acpi_link_adr *link;
-	int num = 0;
+	bool group_visited[MAX_SDW_LINKS];
+	int cpu_num = 0, dai_num = 0;
+	int i;
 
 	link = mach_params->links;
 	if (!link)
 		return -EINVAL;
 
+	for (i = 0; i < MAX_SDW_LINKS; i++)
+		group_visited[i] = FALSE;
+
 	for (; link->num_adr; link++) {
+		const struct snd_soc_acpi_endpoint *endpoints;
 		int part_id, id;
 		u64 adr;
 
-		adr = link->adr_d[0].adr;
+		endpoints = link->adr_d->endpoints;
+		/* only CPU DAI is needed for aggregated link */
+		if (endpoints->aggregated &&
+		    group_visited[endpoints->group_id - 1]) {
+			cpu_num++;
+			continue;
+		}
+
+		adr = link->adr_d->adr;
 
 		part_id = SDW_PART_ID(adr);
 		id = find_codec_info_part(part_id);
 		if (id < 0)
-			return 0;
+			return -EINVAL;
 
 		/* support playback */
-		if (codec_info_list[id].direction[0])
-			num++;
+		if (codec_info_list[id].direction[0]) {
+			cpu_num++;
+			dai_num++;
+		}
 
 		/* support capture */
-		if (codec_info_list[id].direction[1])
-			num++;
+		if (codec_info_list[id].direction[1]) {
+			cpu_num++;
+			dai_num++;
```

Comment:
can we use the new for_each_pcm_stream macro to loop through playback and capture and condense the code here?

Reply:
thanks, I will refine it